### ISSUE TITLE
GTM iRule functionality missing in AS3.

### ISSUE THREAD

**Issue by devildog-tech** [F5Networks/f5-appsvcs-extension#590]
### Is your feature request related to a problem? Please describe.
I have found no reference or ability to create or reference GTM iRules in GSLB_Domain. Unless I am overlooking something. This presents an issue moving to Declarative. Many of our Wide IP's utilize iRules.

### Describe the solution you'd like
Please point me in the direction of how to create and reference GTM iRules if there is the ability or build in the functionality.

### Describe alternatives you've considered
None.

### Additional context
None.

**Comment by dstokesf5** [comment]
Thank you for your feedback. I have added this feature request to our internal product backlog as AUTOTOOL-3129.

**Comment by mdditt2000** [comment]
@devildog-tech are you looking for something like this?

**Provide EDNS/GSLB for both internal and external IPs**

Providing EDNS only for the published IPs (typically internal) only covers a small portion of the use-cases we see. It is important that we can provide DNS resolution for both internal and external IPs. This can be achieved with an iRule that would automatically do the translations based on a DataGroup (potentially). [iRule](https://clouddocs.f5.com/api/irules/GTM-iRule-with-NAT-translation.html)

```
{
    "kind": "tm:gtm:wideip:a:acollectionstate",
    "selfLink": "https://localhost/mgmt/tm/gtm/wideip/a?ver=15.1.4.1",
    "items": [
        {
            "kind": "tm:gtm:wideip:a:astate",
            "name": "www.test.com",
            "partition": "cis-crd",
            "fullPath": "/cis-crd/www.test.com",
            "generation": 1870,
            "selfLink": "https://localhost/mgmt/tm/gtm/wideip/a/~cis-crd~www.test.com?ver=15.1.4.1",
            "enabled": true,
            "failureRcode": "noerror",
            "failureRcodeResponse": "disabled",
            "failureRcodeTtl": 0,
            "lastResortPool": "",
            "minimalResponse": "enabled",
            "persistCidrIpv4": 32,
            "persistCidrIpv6": 128,
            "persistence": "disabled",
            "poolLbMode": "ratio",
            "topologyPreferEdns0ClientSubnet": "enabled",
            "ttlPersistence": 3600,
            "rules": [
                "/cis-crd/test"
            ],
            "rulesReference": [
                {
                    "link": "https://localhost/mgmt/tm/gtm/rule/~cis-crd~test?ver=15.1.4.1"
                }
            ],
            "pools": [
                {
                    "name": "Pool-B",
                    "partition": "cis-crd",
                    "order": 1,
                    "ratio": 1,
                    "nameReference": {
                        "link": "https://localhost/mgmt/tm/gtm/pool/a/~cis-crd~Pool-B?ver=15.1.4.1"
                    }
                },
                {
                    "name": "pool-A",
                    "partition": "cis-crd",
                    "order": 0,
                    "ratio": 1,
                    "nameReference": {
                        "link": "https://localhost/mgmt/tm/gtm/pool/a/~cis-crd~pool-A?ver=15.1.4.1"
                    }
                }
            ]
        }
    ]
}
```

**Comment by devildog-tech** [comment]
This is what I'm looking for. Associating a GSLB iRule to a WIP is only part of the solution. The other half is a new object for GSLB_iRule to be able to create the objects referenced. GSLB_iRule should be more or less a replicate of iRule object as far as functionality is concerned. LTM iRules are seperate from GSLB iRules so there is no object currently in AS3 to be able to create GSLB iRules. Hopefully this helps. 

Note: Not trying to nit pick but I noticed you referenced it as "rules" in your GSLB_Domain object. I would suggest changing that to "iRules" to keep some constancy between LTM and GTM objects. Maybe there is something on the backend that prevents this that I'm not seeing so just a suggestion. for sanity sake.

```
{
    "$schema": "https://raw.githubusercontent.com/F5Networks/f5-appsvcs-extension/master/schema/latest/as3-schema.json",
    "class": "AS3",
    "action": "deploy",
    "persist": true,
    "id": "test.com",
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.0.0",
        "test.com": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "enable": true,
                "test.com_wip_a": {
                    "resourceRecordType": "A",
                    "domainName": "test.com",
                    "iRules":[
                        {
                            "use": "test_irule"
                        },
                        {
                            "bigip": "/Common/Shared/test2_irule"
                        }
                    ],
                    "pools": [
                        {
                            "use": "test.com_a"
                        }
                    ],
                    "enabled": true,
                    "remark": "SCRUBBED",
                    "class": "GSLB_Domain"
                },
                "test.com_wip_aaaa": {
                    "resourceRecordType": "AAAA",
                    "domainName": "test.com",
                    "iRules":[
                        {
                            "use": "test_irule"
                        },
                        {
                            "bigip": "/Common/Shared/test2_irule"
                        }
                    ],
                    "pools": [
                        {
                            "use": "test.com_aaaa"
                        }
                    ],
                    "enabled": false,
                    "remark": "SCRUBBED",
                    "class": "GSLB_Domain"
                },
                "test.com_a": {
                    "resourceRecordType": "A",
                    "ttl": 2,
                    "lbModePreferred": "topology",
                    "lbModeAlternate": "global-availability",
                    "members": [
                        {
                            "virtualServer": "/test.com/Shared/test.com_443_lbve_tcp_vs",
                            "server": {
                                "bigip": "/Common/SCRUBBED"
                            }
                        },
                        {
                            "virtualServer": "/Common/vs_166_123_218_220_https",
                            "server": {
                                "bigip": "/Common/SCRUBBED"
                            }
                        }
                    ],
                    "remark": "SCRUBBED",
                    "class": "GSLB_Pool"
                },
                "test.com_aaaa": {
                    "resourceRecordType": "AAAA",
                    "ttl": 2,
                    "lbModePreferred": "topology",
                    "lbModeAlternate": "global-availability",
                    "members": [
                        {
                            "virtualServer": "/test.com/Shared/test.com_443_lbve_tcp_vs-1-",
                            "server": {
                                "bigip": "/Common/SCRUBBED"
                            }
                        }
                    ],
                    "remark": "SCRUBBED",
                    "class": "GSLB_Pool"
                },
                "test_irule": {
                    "remark": "SCRUBBED",
                    "class": "GSLB_iRule",
                    "iRule": {
                        "url": {
                            "url": "https://SCRUBBED.tcl/raw?ref=main",
                            "authentication": {
                                "method": "bearer-token",
                                "token": "SCRUBBED"
                            }
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by dstokesf5** [comment]
This issue has been resolved with the 3.37.0 release.
